### PR TITLE
Add documentation for the Logfire Query API

### PR DESCRIPTION
**Pull request by dmontagu**
### Query API

Logfire now provides a web API for programmatically running arbitrary SQL queries against the data in your Logfire projects.

This API can be used to retrieve data for export, analysis, or integration with other tools, allowing you to leverage
your data in a variety of ways.

The API is available at `https://logfire-api.pydantic.dev/v1/query` and requires a **read token** for authentication.
You will be soon able to generate read tokens from the Logfire web interface (this is described in more detail in the docs added in this PR).

The API can return data in various formats, including JSON, Apache Arrow, and CSV.

### Changes in this PR:

This PR adds documentation of how to make HTTP requests to this API directly, and also adds sync and async clients in `logfire.experimental.query_client` to make it convenient to write code leveraging these APIs from python.

The clients are being included under the `logfire.experimental` namespace for now to indicate that they may be subject to breaking changes, though we do not have any such changes planned at this time and will try to avoid them as much as possible.

### CODE REVIEW

File: docs/guides/advanced/using_read_tokens.md [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+These clients are currently experimental, meaning we might introduce breaking changes in the future.
+To use these clients, you can import them from the `experimental` namespace:
+
+```python
+from logfire.experimental.read_client import LogfireAsyncReadClient, LogfireSyncReadClient
+```
+
+### Async Client Example
+
+The `LogfireAsyncReadClient` allows for asynchronous interaction with the Logfire API. Here's an example of how to use it:
```
Comment (Kludex) [inline docs/guides/advanced/using_read_tokens.md]: I think it should be called `AsyncLogfireClient` or `AsyncLogfireReadClient`. It's more common in the community.

File: docs/guides/advanced/using_read_tokens.md [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+        # Retrieve data in CSV format, and load into a polars DataFrame
+        # Note that JSON columns such as `attributes` will be returned as JSON-serialized strings
+        df_from_csv = pl.read_csv(StringIO(await client.query_csv(sql=query)))
+        print(df_from_csv)
+
+
+asyncio.run(main())
+```
+
+### Sync Client Example
```
Comment (Kludex) [inline docs/guides/advanced/using_read_tokens.md]: I don't think there's need for the same example but sync, you can use the tab in mkdocs-material notation.

Also, I think this client should be called `LogfireReadClient` or LogfireClient`, without the `Sync`.

File: logfire/experimental/read_client.py [hunk 3]
```diff
@@ -0,0 +1,338 @@
+from __future__ import annotations
+
+from datetime import datetime
+from types import TracebackType
+from typing import TYPE_CHECKING, Annotated, Any, Generic, Literal, Self, TypedDict, TypeVar
+
+from annotated_types import Gt, Le
```
Comment (alexmojaki) [inline logfire/experimental/read_client.py]: This isn't in the normal dependencies, either users have to install it or it should only go under TYPE_CHECKING if possible.

**Comment by Kludex** [comment]
Should I block this bot from the organization?

File: docs/guides/advanced/query_api.md [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+
+The API is available at `https://logfire-api.pydantic.dev/v1/query` and requires a **read token** for authentication.
+Read tokens can be generated from the Logfire web interface and provide secure access to your data.
+
+The API can return data in various formats, including JSON, Apache Arrow, and CSV, to suit your needs.
+See [here](#additional-configuration) for more details about the available response formats.
+
+## How to Create a Read Token
+
+If you've set up Logfire following the [first steps guide](../first_steps/index.md), you can generate read tokens from
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply docs/guides/advanced/query_api.md:13]: Yes

File: docs/guides/advanced/query_api.md [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+the Logfire web interface, for use accessing the Logfire Query API.
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply docs/guides/advanced/query_api.md:14]: Yes

File: docs/guides/advanced/query_api.md [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+
+To create a read token:
+
+1. Open the **Logfire** web interface at [logfire.pydantic.dev](https://logfire.pydantic.dev).
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply docs/guides/advanced/query_api.md:18]: Yes

File: .hyperlint/styles/config/vocabularies/hyperlint/accept.txt [hunk 7]
```diff
@@ -0,0 +1,3 @@
+Logfire
+namespace
+Async
```
Comment (bllchmbrs) [inline .hyperlint/styles/config/vocabularies/hyperlint/accept.txt]: Is it ok if async is lowercase? You can do a regex here too `[aA]sync`.

Comment (bllchmbrs) [reply .hyperlint/styles/config/vocabularies/hyperlint/accept.txt]: I would also add:

```
validator
Pydantic
validators
namespace
Hyperlint
preprocess
tokenization
tokenizer
API
APIs
SDKs
SDK
```

They're going to come up.

File: docs/guides/advanced/query_api.md [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+The Logfire API supports various query parameters and response formats to give you flexibility in how you retrieve your data:
+
+- **Response Format**: Use the `Accept` header to specify the response format. Supported values include:
+  - `application/json`: Returns the data in JSON format. By default, this will be column-oriented unless specified otherwise with the `json_rows` parameter.
+  - `application/vnd.apache.arrow.stream`: Returns the data in Apache Arrow format, suitable for high-performance data processing.
+  - `text/csv`: Returns the data in CSV format, which is easy to use with many data tools.
+
+  If no `Accept` header is provided, the default response format is JSON.
+
+- **Query Parameters**:
```
Comment (samuelcolvin) [inline docs/guides/advanced/query_api.md:209]: this doesn't render properly

Comment (samuelcolvin) [reply docs/guides/advanced/query_api.md:209]: and `sql` needs to be included.

File: docs/guides/advanced/query_api.md [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+import requests
+
+# Define the base URL and your read token
+base_url = 'https://logfire-api.pydantic.dev'
+read_token = '<your_read_token_here>'
+
+# Set the headers for authentication
+headers = {
+    'Authorization': f'Bearer {read_token}',
+    'Content-Type': 'application/json'
```
Comment (samuelcolvin) [inline docs/guides/advanced/query_api.md:171]: This needs to be `Accept`, not `Content-Type`

File: docs/guides/advanced/query_api.md [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+### Additional Configuration
+
+The Logfire API supports various query parameters and response formats to give you flexibility in how you retrieve your data:
+
+- **Response Format**: Use the `Accept` header to specify the response format. Supported values include:
+  - `application/json`: Returns the data in JSON format. By default, this will be column-oriented unless specified otherwise with the `json_rows` parameter.
+  - `application/vnd.apache.arrow.stream`: Returns the data in Apache Arrow format, suitable for high-performance data processing.
+  - `text/csv`: Returns the data in CSV format, which is easy to use with many data tools.
+
+  If no `Accept` header is provided, the default response format is JSON.
```
Comment (samuelcolvin) [inline docs/guides/advanced/query_api.md:207]: It seems to be required.

**Pull request merged**